src/webgpu/util/math.ts

```diff
@@ -295,3 +307,16 @@ export function biasedRange(min: Scalar, max: Scalar, num_steps: Scalar): Array<
     lerp(f32_min, f32_max, Math.pow(lerp(0, 1, i / (u32_num_steps - 1)), c))
   );
 }
+
+/** Generates all of the pairs created by taking the cartesian product of an array with itself. */
+export function cartesianPairs<T>(values: T[]): [T, T][] {
+  const result = new Array<[T, T]>();
+  for (let lhs_index = 0; lhs_index < values.length; lhs_index++) {
```

Comment:
I think you could do this instead of the nested loop:

```js
values.forEach((e,ei) => {
  values.forEach((f,fi) => {
    result.push([e,f]);
  })
});
```

Reply:
whoops. In this part of the code you don't need the ei and fi arguments...
(And you may not need this function at all)